### PR TITLE
build: add script to publish release

### LINKED ISSUE #13455: Improve release process automation
#### Create a "stage release" task that:
- [x] Prompts for which new version (major, minor, patch, noop, custom)
- [x] Asserts that the version is the correct increment from an existing version
- [x] Asserts that user is on the correct branch
- [x] Asserts no local changes (except for the version if "noop" version bump)
- [x] Asserts that branch is up-to-date with `upstream`
- [x] (Bonus) Assert that CI is green for that branch
- [x] Switches to a new release branch (e.g. `release-6.4.2`)
- [x] Generates the changelog
- [x] Commits the version bump and changelog (`chore: bump version to X w/ changelog`)
- [ ] (Bonus) Creates a PR

#### The publish task should (in addition to what it already does):
- [x] Check if last commit is a version bump
- [x] Check publish branch
- [x] Check if local head = remote head
- [x] Prompt for NPM dist-tag & validate (depending on pre-release)
   - [x] For a non-prerelease version that selects "next", show an additional prompt to confirm
- [x] Build release output
- [x] Run release output validations
- [x] Tag release w/ changelog
  - [ ] Pull in changelog
- [x] Interactively run `npm login` if not logged in
- [x] run "npm publish" for each package
- [ ] Push tag to upstream?
- [ ] (Bonus) Prompt a warning if master is red on CI

#### The changelog task should:
- [x] Prompt for a release name (stable releases only) and include it in the changelog

#### Additional TODO's

```
  // TODO(devversion): minor releases can also be published from master if there isn't a minor

```

cc @josephperrott for more ideas

### PR DESCRIPTION
**Pull request by devversion**
Closes #13455

### CODE REVIEW

File: tools/release/git-release-task.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import {green, italic, red, yellow} from 'chalk';
+import {GitClient} from './git/git-client';
+import {Version} from './version-name/parse-version';
+import {getAllowedPublishBranches} from './version-name/publish-branches';
+
+export class GitReleaseTask {
```
Comment (jelbourn) [inline tools/release/git-release-task.ts]: Add a class description?

File: tools/release/publish-release.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    // the issues need to be resolved before publishing.
+    if (hasFailed) {
+      console.error(red(`  ✘   Release output does not pass all release validations. Please fix ` +
+        `all failures or reach out to the team.`));
+      process.exit(1);
+    }
+  }
+
+  /**
+   * Prompts the user whether he is sure that the current stable version should be
```
Comment (jelbourn) [inline tools/release/publish-release.ts]: "he is" -> "they are"

Comment (devversion) [reply tools/release/publish-release.ts]: Done.

File: tools/release/git-release-task.ts (hunk 1 above)
Comment (devversion) [reply tools/release/git-release-task.ts]: Done.

**Comment by devversion** [comment]
@jelbourn I've added the remaining changes I planned to add to this PR. Please have another look.

**Pull request merged**